src/appleseed/foundation/meta/tests/test_filtersamplingtable.cpp

```diff
@@ -0,0 +1,207 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.foundation headers.
+#include "foundation/math/cdf.h"
+#include "foundation/math/filter.h"
+#include "foundation/math/filtersamplingtable.h"
+#include "foundation/math/rng/distribution.h"
+#include "foundation/math/rng/xoroshiro128plus.h"
+#include "foundation/utility/test.h"
+
+// Standard headers.
+#include <cstddef>
+#include <cstdio>
+#include <fstream>
+
+using namespace foundation;
+using namespace std;
+
+namespace
+{
+    template <typename Filter1, typename Filter2>
+    void generate_sample_plot_compare(
+        const Filter1&  filter1,
+        const Filter2&  filter2,
+        const size_t    sample_count,
+        const char*     filename)
+    {
+        ofstream ofs(filename);
+        Xoroshiro128plus rng(349, 684658);
+
+        // Generate samples using the table.
+        FilterSamplingTable table(filter1);
+
+        ofs << "points1 = [\n";
+
+        for (size_t i = 0; i < sample_count; ++i)
+        {
+            const float sx = rand_float1(rng);
+            const float sy = rand_float1(rng);
+            const Vector2f p(table.sample(sx), table.sample(sy));
+
+            if (p.x < 0.0f)
+            {
+                ofs << "(" << p.x << ", " << p.y << ")";
+
+                if (i != sample_count - 1)
+                    ofs << ",";
+
+                ofs << "\n";
+            }
+        }
+
+        ofs << "]\n";
+        ofs << "\n";
+
+        // Generate samples using the 2D filter and rejection sampling.
+        const float rx = filter2.get_xradius();
+        const float ry = filter2.get_yradius();
+        const float rcp_norm = 1.0f / compute_normalization_factor(filter2);
+
+        ofs << "points2 = [\n";
+
+        for (size_t i = 0; i < sample_count; ++i)
+        {
+            while (true)
+            {
+                const float sx = rand_float1(rng);
+                const float sy = rand_float1(rng);
+
+                const float x = lerp(-rx, rx, sx);
+                const float y = lerp(-ry, ry, sy);
+
+                const float value = filter2.evaluate(x, y) * rcp_norm;
+
+                if (value > rand_float1(rng))
+                {
+                    if (x > 0.0f)
+                    {
+                        ofs << "(" << x << ", " << y << ")";
+
+                        if (i != sample_count - 1)
+                            ofs << ",";
+
+                        ofs << "\n";
+                        break;
+                    }
+                }
+            }
+        }
+
+        ofs << "]\n";
+        ofs << "\n";
+
+        ofs << "import matplotlib.pyplot as plt\n";
```

Comment:
Can't we refactor this ? Ploting stuff could be useful in a lot of test

Reply:
I guess it's fine having this code here for now. We could promote it to foundation when a second use case arises.

Note that we already have somewhat generic plotting facilities based on GnuPlot, and many "unit" tests actually generate plots.